File: apps/api/app/schemas/nutrition.py
"""Schemas for the nutrition endpoints.

Covers:
  Food          CRUD (create custom food, list/search, get, update, delete)
  FoodLog       log a food entry, list by date, update, delete
  WaterLog      log water, list by date, update, delete
  DailyNutrition  aggregated daily summary (totals + per-meal breakdown)
"""

from __future__ import annotations

import uuid
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field

# ── Meal type ─────────────────────────────────────────────────────────────────

MealType = Literal["breakfast", "lunch", "dinner", "snack", "other"]

MEAL_TYPE_ORDER: list[MealType] = ["breakfast", "lunch", "dinner", "snack", "other"]


# ── Food ──────────────────────────────────────────────────────────────────────


class CreateFoodRequest(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    brand: str | None = Field(default=None, max_length=200)
    description: str | None = Field(default=None, max_length=1000)

    # All macros per 100 g
    calories_per_100g: float = Field(ge=0)
    protein_per_100g: float = Field(default=0.0, ge=0)
    carbs_per_100g: float = Field(default=0.0, ge=0)
    fat_per_100g: float = Field(default=0.0, ge=0)
    fiber_per_100g: float | None = Field(default=None, ge=0)
    sugar_per_100g: float | None = Field(default=None, ge=0)
    sodium_per_100g: float | None = Field(default=None, ge=0)

    serving_size_g: float | None = Field(default=None, gt=0)
    serving_unit: str | None = Field(default=None, max_length=50)


class UpdateFoodRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    brand: str | None = None
    description: str | None = None
    calories_per_100g: float | None = Field(default=None, ge=0)
    protein_per_100g: float | None = Field(default=None, ge=0)
    carbs_per_100g: float | None = Field(default=None, ge=0)
    fat_per_100g: float | None = Field(default=None, ge=0)
    fiber_per_100g: float | None = Field(default=None, ge=0)
    sugar_per_100g: float | None = Field(default=None, ge=0)
    sodium_per_100g: float | None = Field(default=None, ge=0)
    serving_size_g: float | None = Field(default=None, gt=0)
    serving_unit: str | None = None


class FoodResponse(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID | None
    name: str
    brand: str | None
    description: str | None
    calories_per_100g: float
    protein_per_100g: float
    carbs_per_100g: float
    fat_per_100g: float
    fiber_per_100g: float | None
    sugar_per_100g: float | None
    sodium_per_100g: float | None
    serving_size_g: float | None
    serving_unit: str | None
    is_system: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class FoodListResponse(BaseModel):
    foods: list[FoodResponse]
    total: int
    page: int
    page_size: int


# ── FoodLog ───────────────────────────────────────────────────────────────────


class LogFoodRequest(BaseModel):
    food_id: uuid.UUID
    logged_date: date
    meal_type: MealType = "other"
    quantity_g: float = Field(gt=0)
    notes: str | None = Field(default=None, max_length=500)


class UpdateFoodLogRequest(BaseModel):
    meal_type: MealType | None = None
    quantity_g: float | None = Field(default=None, gt=0)
    notes: str | None = None


class FoodLogResponse(BaseModel):
    id: uuid.UUID
    food_id: uuid.UUID
    logged_date: date
    meal_type: str
    quantity_g: float
    notes: str | None

    # Computed macros for this log entry (scaled from per-100g values)
    calories: float
    protein_g: float
    carbs_g: float
    fat_g: float
    fiber_g: float | None

    # Denormalised food info for display
    food_name: str
    food_brand: str | None

    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ── WaterLog ──────────────────────────────────────────────────────────────────


class LogWaterRequest(BaseModel):
    logged_date: date
    amount_ml: int = Field(gt=0, le=10_000)
    notes: str | None = Field(default=None, max_length=500)


class UpdateWaterLogRequest(BaseModel):
    amount_ml: int | None = Field(default=None, gt=0, le=10_000)
    notes: str | None = None


class WaterLogResponse(BaseModel):
    id: uuid.UUID
    logged_date: date
    amount_ml: int
    notes: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


# ── Daily summary ─────────────────────────────────────────────────────────────


class MacroTotals(BaseModel):
    """Aggregated macro totals for a meal slot or the whole day."""

    calories: float
    protein_g: float
    carbs_g: float
    fat_g: float
    fiber_g: float


class MealSection(BaseModel):
    meal_type: str
    entries: list[FoodLogResponse]
    totals: MacroTotals


class DailyNutritionResponse(BaseModel):
    """Full nutrition picture for a single date."""

    date: date
    meals: list[MealSection]  # one per meal_type that has at least one entry
    day_totals: MacroTotals
    water_logs: list[WaterLogResponse]
    water_total_ml: int

File: apps/web/src/types/nutrition.ts
/**
 * TypeScript types for the FitTrack AI nutrition domain.
 *
 * Mirrors the Pydantic schemas in apps/api/app/schemas/nutrition.py
 */

// ── Shared ─────────────────────────────────────────────────────────────────────

export type MealType = "breakfast" | "lunch" | "dinner" | "snack" | "other";

export const MEAL_TYPE_LABELS: Record<MealType, string> = {
  breakfast: "Breakfast",
  lunch: "Lunch",
  dinner: "Dinner",
  snack: "Snack",
  other: "Other",
};

export const MEAL_TYPE_ORDER: MealType[] = [
  "breakfast",
  "lunch",
  "dinner",
  "snack",
  "other",
];

// ── Food ───────────────────────────────────────────────────────────────────────

export interface Food {
  id: string;
  user_id: string | null;
  name: string;
  brand: string | null;
  description: string | null;
  calories_per_100g: number;
  protein_per_100g: number;
  carbs_per_100g: number;
  fat_per_100g: number;
  fiber_per_100g: number | null;
  sugar_per_100g: number | null;
  sodium_per_100g: number | null;
  serving_size_g: number | null;
  serving_unit: string | null;
  is_system: boolean;
  created_at: string;
  updated_at: string;
}

export interface FoodListResponse {
  foods: Food[];
  total: number;
  page: number;
  page_size: number;
}

export interface CreateFoodPayload {
  name: string;
  brand?: string;
  description?: string;
  calories_per_100g: number;
  protein_per_100g?: number;
  carbs_per_100g?: number;
  fat_per_100g?: number;
  fiber_per_100g?: number;
  serving_size_g?: number;
  serving_unit?: string;
}

export interface MacroPortion {
  grams: number;
  calories_kcal: number;
  protein_g: number;
  carbs_g: number;
  fat_g: number;
  fiber_g: number | null;
}

/** AI macro estimate - a preview the user edits before saving. */
export interface MacroEstimate {
  ai_available: boolean;
  name: string | null;
  serving_size_g: number | null;
  serving_unit: string | null;
  calories_per_100g: number | null;
  protein_per_100g: number | null;
  carbs_per_100g: number | null;
  fat_per_100g: number | null;
  fiber_per_100g: number | null;
  portion: MacroPortion | null;
  confidence: string | null;
  is_estimate: boolean;
  disclaimer: string;
  message: string | null;
  provider: string | null;
  model_id: string | null;
  prompt_version: string | null;
  log_id: string | null;
}

// ── Multi-item meal estimate (several foods described at once) ─────────────────

export interface MealItemEstimate {
  name: string;
  quantity_g: number;
  serving_unit: string | null;
  calories_per_100g: number;
  protein_per_100g: number;
  carbs_per_100g: number;
  fat_per_100g: number;
  fiber_per_100g: number | null;
  confidence: string;
  portion: MacroPortion;
}

/** AI estimate for a whole meal description - a preview, editable per item. */
export interface MealEstimate {
  ai_available: boolean;
  items: MealItemEstimate[];
  totals: MacroPortion | null;
  is_estimate: boolean;
  disclaimer: string;
  message: string | null;
  provider: string | null;
  model_id: string | null;
  prompt_version: string | null;
  log_id: string | null;
}

export interface LogMealItemPayload {
  name: string;
  quantity_g: number;
  serving_unit?: string;
  calories_per_100g: number;
  protein_per_100g?: number;
  carbs_per_100g?: number;
  fat_per_100g?: number;
  fiber_per_100g?: number;
}

export interface LogMealPayload {
  logged_date: string; // YYYY-MM-DD
  meal_type?: MealType;
  items: LogMealItemPayload[];
  estimate_log_id?: string;
}

export interface LogMealResult {
  entries: FoodLogEntry[];
  totals: MacroTotals;
}

// ── Nutrition targets (user-configured daily goals) ─────────────────────────────

export interface NutritionTarget {
  calorie_target_kcal: number | null;
  protein_target_g: number | null;
  carbs_target_g: number | null;
  fat_target_g: number | null;
  fiber_target_g: number | null;
  is_set: boolean;
  updated_at: string | null;
}

export interface UpdateNutritionTargetPayload {
  calorie_target_kcal?: number | null;
  protein_target_g?: number | null;
  carbs_target_g?: number | null;
  fat_target_g?: number | null;
  fiber_target_g?: number | null;
}

// ── Daily nutrition insight (read-only AI comparison + suggestions) ────────────

export interface MacroComparison {
  metric: "calories" | "protein" | "carbs" | "fat" | "fiber";
  label: string;
  unit: string;
  current: number;
  target: number | null;
  percent_of_target: number | null;
  remaining: number | null;
}

export interface DailyInsight {
  date: string;
  day_totals: MacroTotals;
  targets: NutritionTarget;
  comparisons: MacroComparison[];
  meals_logged: string[];
  meals_remaining: string[];
  ai_available: boolean;
  highlights: string[];
  suggestions: string[];
  encouragement: string;
  disclaimer: string;
  message: string | null;
  provider: string | null;
  model_id: string | null;
  prompt_version: string | null;
  log_id: string | null;
  generated_at: string | null;
}

export interface UpdateFoodPayload {
  name?: string;
  brand?: string;
  description?: string;
  calories_per_100g?: number;
  protein_per_100g?: number;
  carbs_per_100g?: number;
  fat_per_100g?: number;
  fiber_per_100g?: number;
  serving_size_g?: number;
  serving_unit?: string;
}

// ── Food log ──────────────────────────────────────────────────────────────────

export interface FoodLogEntry {
  id: string;
  food_id: string;
  logged_date: string;
  meal_type: string;
  quantity_g: number;
  notes: string | null;
  // computed macros for this entry
  calories: number;
  protein_g: number;
  carbs_g: number;
  fat_g: number;
  fiber_g: number | null;
  // denormalised food info
  food_name: string;
  food_brand: string | null;
  created_at: string;
  updated_at: string;
}

export interface LogFoodPayload {
  food_id: string;
  logged_date: string; // YYYY-MM-DD
  meal_type?: MealType;
  quantity_g: number;
  notes?: string;
}

export interface UpdateFoodLogPayload {
  meal_type?: MealType;
  quantity_g?: number;
  notes?: string;
}

// ── Water log ─────────────────────────────────────────────────────────────────

export interface WaterLogEntry {
  id: string;
  logged_date: string;
  amount_ml: number;
  notes: string | null;
  created_at: string;
}

export interface LogWaterPayload {
  logged_date: string; // YYYY-MM-DD
  amount_ml: number;
  notes?: string;
}

export interface UpdateWaterLogPayload {
  amount_ml?: number;
  notes?: string;
}

// ── Daily summary ─────────────────────────────────────────────────────────────

export interface MacroTotals {
  calories: number;
  protein_g: number;
  carbs_g: number;
  fat_g: number;
  fiber_g: number;
}

export interface MealSection {
  meal_type: string;
  entries: FoodLogEntry[];
  totals: MacroTotals;
}

export interface DailyNutrition {
  date: string;
  meals: MealSection[];
  day_totals: MacroTotals;
  water_logs: WaterLogEntry[];
  water_total_ml: number;
}
